src/Orleans.Runtime/Catalog/ActivationCollector.cs

```diff
@@ -302,6 +319,103 @@ public List<ICollectibleGrainContext> ScanAll(TimeSpan ageLimit)
         }
 
+        /// <summary>
+        /// Checks if current memory usage is above the configured threshold for deactivation.
+        /// Also calculates the target number of activations to deactivate if memory is overloaded to reach target memory usage.
+        /// </summary>
+        /// <remarks>internal for testing</remarks>
+        /// <param name="targetDeactivationNumber">number of deactivations to perform to bring back memory consumption to target value</param>
+        internal bool IsMemoryOverloaded(out int targetDeactivationNumber)
+        {
+            targetDeactivationNumber = 0;
+            var stats = _environmentStatisticsProvider.GetEnvironmentStatistics();
+
+            var usedMemory = stats.MaximumAvailableMemoryBytes - stats.AvailableMemoryBytes;
+            var memoryLoadPercentage = 100.0 * usedMemory / stats.MaximumAvailableMemoryBytes;
+            var threshold = _memoryBasedGrainCollectionOptions.MemoryLoadThresholdPercentage;
+            var targetThreshold = _memoryBasedGrainCollectionOptions.TargetMemoryLoadPercentage;
+
+            if (memoryLoadPercentage < threshold)
+            {
+                return false;
+            }
+
+            var targetUsedMemory = targetThreshold * stats.MaximumAvailableMemoryBytes / 100.0;
+            var memoryToFree = usedMemory - targetUsedMemory;
+
+            // Can't know for sure, but overestimating here
+            var activationSize = usedMemory / (double)_activationCount;
+            targetDeactivationNumber = (int)Math.Ceiling(memoryToFree / activationSize);
+            if (targetDeactivationNumber < 1)
+            {
+                targetDeactivationNumber = 1;
+            }
+
+            return true;
+        }
+
+        /// <summary>
+        /// Deactivates <param name="deactivationsNumber" /> activations in due time order
+        /// <remarks>internal for testing</remarks>
+        /// </summary>
+        internal async Task DeactivateInDueTimeOrder(int deactivationsNumber, CancellationToken cancellationToken)
+        {
+            if (deactivationsNumber <= 0)
+            {
+                LogNoDeactivationsFound();
+                return;
+            }
+
+            var watch = ValueStopwatch.StartNew();
+            var number = Interlocked.Increment(ref collectionNumber);
+            long memBefore = GC.GetTotalMemory(false) / (1024 * 1024);
+            LogBeforeCollection(number, memBefore, _activationCount, this);
+
+            var candidates = new List<ICollectibleGrainContext>(deactivationsNumber);
+            var selectedCount = 0;
+
+            foreach (var bucket in buckets.OrderBy(b => b.Key))
+            {
+                foreach (var item in bucket.Value.Items)
+                {
+                    var activation = item.Value;
+                    lock (activation)
+                    {
+                        // TODO - do we need to check some properties of activation here?
```

Comment:
@ReubenBond what do u think about validation here? Should we skip some activations?